app/services/versioned_files_service/metadata.rb

```diff
@@ -12,7 +12,32 @@ def initialize(paths:)
     def write_cocina(version:, cocina:)
       FileUtils.mkdir_p(versions_path)
       cocina_path = cocina_path_for(version:)
-      cocina_path.write(cocina.to_json)
+      cocina_path.write(self.class.deep_compact_hash(cocina.to_h).to_json)
+    end
+
+    def self.deep_compact_hash(hash) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+      hash.each_with_object({}) do |(key, value), new_hash|
+        case value
+        when Hash
+          compacted_value = deep_compact_hash(value)
+          new_hash[key] = compacted_value unless compacted_value.empty?
+        when Array
+          # Compact arrays by removing nils and empty hashes/arrays within them
+          compacted_array = value.filter_map do |item|
+            if item.is_a?(Hash)
+              deep_compact_hash(item)
+            elsif item.is_a?(Array)
+              # Recursively deep compact nested arrays if needed
+              item.filter_map { |sub_item| sub_item.is_a?(Hash) ? deep_compact_hash(sub_item) : sub_item }.reject(&:empty?)
+            else
+              item
+            end
```

Comment:
Do we have to do these checks here, rather than just letting them happen when we recurse? The [cocina_display implementation](https://github.com/sul-dlss/cocina_display/blob/d0118d9e73a15640d07a9f889e0cc35ce55bd507/lib/cocina_display/utils.rb#L53-L77) ended up less verbose. Maybe I missed something, though?